### PR TITLE
Improved administration page for refreshing AniDB tags

### LINKED ISSUE #205: Refaire la page d'admin  pour la modification des tags avec un Form Django
Refaire la page d'admin pour la modification des tags avec un formulaire Form Django : voir le FIXME dans admin.py sur la branche fravoi/anidb

### PR DESCRIPTION
**Pull request by mrsalixor**
Closes #205 by improving the administration page for refreshing AniDB tags :
- Cleaner disposition
- Possibility to **modify the weight of tags** that were modified, added or kept by AniDB
- Possibility to **choose which works should have their tags updated** (after reviewing new tag information from AniDB)
- Possibility to **select which tags should be added, updated, kept or deleted** (by default, the checkboxes are checked meaning the planned operation will happen when submitting, given the work is marked for update)

Also moves the `Work.retrieve_tags` method to `utils/anidb/diff_between_anidb_and_local_tags`.
And no longer calls the API twice in order to only update 1 work !

### CODE REVIEW

File: mangaki/mangaki/admin.py [hunk 1]
```diff
@@ -229,60 +230,103 @@ def make_nsfw(self, request, queryset):
...
-
-                client.update_tags(anime, tags)
+        works = queryset.all()
 
-            self.message_user(request, "Modifications sur les tags faites")
+        if request.POST.get('confirm'): # Updating tags has been confirmed
+            to_update_work_ids = set(list(map(int, request.POST.getlist('to_update_work_ids'))))
+            nb_updates = len(to_update_work_ids)
+
+            work_ids = list(map(int, request.POST.getlist('work_ids')))
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: AFAIK, `list` is unnecessary here because you only want to iterate on the data, not to store it somewhere.

File: mangaki/mangaki/admin.py [hunk 2]
```diff
@@ -229,60 +230,103 @@ def make_nsfw(self, request, queryset):
...
+            tag_anidb_tag_ids = list(map(int, request.POST.getlist('anidb_tag_ids')))
+            tag_operations = request.POST.getlist('tag_operations')
+
+            # These checkboxes make it possible to know which tags should be added, kept, modified or removed
+            tag_checkboxes = request.POST.getlist('tag_checkboxes')
+            tags_to_process = [list(map(int, tag_checkbox.split(':'))) for tag_checkbox in tag_checkboxes]
+
+            # Build a dict that links a work ID and the dict of tags
+            final_tags_with_work_id = {}
+            for index, work_id in enumerate(work_ids):
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Is there any reason to not iterate over `tags_to_process` which would save a `O(N)` lookup where `N = len(tags_to_process)` for each `work_id`?
If it's because you need the `index` to access structured data, I'd suggest to (in any order):
1) recombine the different tag_stuff into an tags with a Tag (namedtuple or whatever), make dicts where you need.
2) use a `dict` for `tags_to_process` (use as key, something like `{{work_id}}_{{anidb_tag_id}}`).
3) it's actually okay to do two loops to process "to delete works" and "to add/modify/keep works" if that simplify your logic / improve the performance.

File: mangaki/mangaki/admin.py [hunk 3]
```diff
@@ -229,60 +230,103 @@ def make_nsfw(self, request, queryset):
...
+                process = [work_id, tag_anidb_tag_ids[index]] in tags_to_process
+                to_delete = tag_operations[index] == 'deleted'
+                to_add_or_update_or_keep = tag_operations[index] in ('added', 'modified', 'kept')
+
+                if not final_tags_with_work_id.get(work_id):
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: That could be written as `work_id not in final_tags_with_work_id`

File: mangaki/mangaki/admin.py [hunk 4]
```diff
@@ -229,60 +230,103 @@ def make_nsfw(self, request, queryset):
...
+                    final_tags_with_work_id[work_id] = {}
+
+                # Only add this tag to the list of tags to process if it has been checked
+                if (process and to_add_or_update_or_keep) or (not process and to_delete):
+                    final_tags_with_work_id[work_id].update({
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Don't use `update` for a single record addition, prefer:
```python
final_tags_with_work_id[work_id][tag_titles[index]] = { stuff }
```
(which is verbose, indeed, but save a `dict` object instantiation and all the stuff which follows).

File: mangaki/mangaki/admin.py [hunk 5]
```diff
@@ -229,60 +230,103 @@ def make_nsfw(self, request, queryset):
...
+                if work.id in to_update_work_ids:
+                    client.update_tags(work, final_tags_with_work_id[work.id])
+
+            if nb_updates == 0:
+                self.message_user(request,
+                                  "Aucune oeuvre n'a été marquée comme devant être mise à jour.",
+                                  level=messages.WARNING)
+            elif nb_updates == 1:
+                self.message_user(request,
+                                  "Mise à jour des tags effectuée pour 1 œuvre.")
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Nitpick: You can even put  "une" rather than "1", seems cheap (even though I know it's done everywhere else haha).

File: mangaki/mangaki/admin.py [hunk 6]
```diff
@@ -229,60 +230,103 @@ def make_nsfw(self, request, queryset):
...
+            # These checkboxes make it possible to know which tags should be added, kept, modified or removed
+            tag_checkboxes = request.POST.getlist('tag_checkboxes')
+            tags_to_process = [list(map(int, tag_checkbox.split(':'))) for tag_checkbox in tag_checkboxes]
+
+            # Build a dict that links a work ID and the dict of tags
+            final_tags_with_work_id = {}
+            for index, work_id in enumerate(work_ids):
+                process = [work_id, tag_anidb_tag_ids[index]] in tags_to_process
+                to_delete = tag_operations[index] == 'deleted'
+                to_add_or_update_or_keep = tag_operations[index] in ('added', 'modified', 'kept')
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: **Micro-optimization / Nice to know**: You could simplify the evaluation using `to_delete`, if you're afraid of bogus operations, you can join the same condition through `and` and still benefit from short-circuit in the case of `to_delete == True`.
[Python documentation about short-circuit operators](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not).

**Another language example of short-circuiting**:
It relies on the same principle in C/C++:
```cpp
bool is_sleeping = is_he_really_sleeping(&raito);
if (is_sleeping || is_playing(&raito)) {
}
```
Assuming `is_playing` is an expensive function, if `is_sleeping = true`, `is_playing` won't be evaluated.

File: mangaki/mangaki/tests/test_refresh_from_anidb.py [hunk 7]
```diff
@@ -87,6 +87,6 @@ def test_refresh_tags_from_anidb_confirmed(self):
         }
 
         response = self.client.post(refresh_tags_from_anidb_url, context)
-        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response.status_code, 200)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_refresh_from_anidb.py]: I'm not sure, why this change?

File: mangaki/mangaki/utils/anidb.py [hunk 8]
```diff
@@ -502,3 +502,38 @@ def get_or_update_work(self,
     getattr(settings, 'ANIDB_CLIENT', None),
     getattr(settings, 'ANIDB_VERSION', None)
 )
+
+def diff_between_anidb_and_local_tags(work: Work,
+                                      anidb_tags: Dict[str, Dict[str, Any]]) -> Dict[str, Dict[str, Dict[str, Any]]]:
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: You can write type alias to improve readability, if you feel like:
```python
Tag = Dict[str, Any]

blablabla
-> Dict[str, Dict[str, Tag]]
```

But IMHO, at this point, we may enjoy the cost of a `typing.NamedTuple`.

File: mangaki/mangaki/tests/test_refresh_from_anidb.py (hunk 7 above)
Comment (mrsalixor) [inline mangaki/mangaki/tests/test_refresh_from_anidb.py]: My bad ! I switched from `302` to `200` since this test failed and thought it came from modifications in my code. But looking back, the return code does not change.

Now, I found out the real cause of this test failing : I didn't update the `context` for the test, and especially `post` was modified to `confirm` 😅
This is now fixed !

File: mangaki/mangaki/admin.py (hunk 5 above)
Comment (mrsalixor) [inline mangaki/mangaki/admin.py]: Indeed, it's better to use text rather than a number here 😄

File: mangaki/mangaki/admin.py (hunk 2 above)
Comment (mrsalixor) [inline mangaki/mangaki/admin.py]: I'll look into this, and also try to improve how AniDB tags are stored (my current `{tag_title: {tag_weight: XX, anidb_tag_id: XX}}` representation is kind of complicated)

**Comment by mrsalixor** [comment]
I made some changes, mostly around how I handled tags from AniDB. I went from a `{tag_title: {'anidb_tag_id': id, 'weight': w}}` to a `namedtuple` called `AniDBTag` in the API wrapper. All the changes that had to be made have been made accordingly.

I took that opportunity to slightly clean the `update_tags` method in the API wrapper by calling `diff_between_anidb_and_local_tags` instead of repeating the same code twice.

Also, it wasn't necessary to check by default tags that are 'deleted' (instead, one just has to check a work if they want to keep it anyway - and we default to unchecked), which removes the need for that `tag_operations` list.

File: mangaki/mangaki/admin.py [hunk 9]
```diff
@@ -229,60 +230,99 @@ def make_nsfw(self, request, queryset):
...
+            tag_titles = request.POST.getlist('tag_titles')
+            tag_weights = list(map(int, request.POST.getlist('weights')))
+            tag_anidb_tag_ids = list(map(int, request.POST.getlist('anidb_tag_ids')))
+            tags = list(map(AniDBTag, tag_titles, tag_weights, tag_anidb_tag_ids))
 
-            self.message_user(request, "Modifications sur les tags faites")
-            return None
+            # Checkboxes to know which tags have to be kept regardless of their pending status
+            tag_checkboxes = request.POST.getlist('tag_checkboxes')
+            tags_to_process = [tuple(map(int, tag_checkbox.split(':'))) for tag_checkbox in tag_checkboxes]
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: What about this instead?
```
tags_to_process = [tuple(map(int, tag_checkbox.split(':')))) for tag_checkbox in tag_checkboxes]
tags_to_process = {'{},{}'.format(work_id, tag_aid): (work_id, tag_aid) for work_id, tag_aid in tags_to_process}
```
Resulting in `O(N)` construction of the dict where `N = len(tag_checkboxes)`.

Then, checking if a couple `(work_id, tag_aid)` is to be processed can be done:
`('{},{}'.format(work_id, tag_aid)) in tags_to_process` and results in a `O(1)` lookup.

Resulting in a total algorithm (starting from line 253) of `O(len(work_ids))` rather than `O(len(work_ids) * len(tags_to_process))` (due to line 258 which is a `O(len(tags_to_process))` lookup).

Of course, using hash maps / dicts is not always a solution, sometimes, you'll have to do tradeoffs on memory consumption versus lookup time and it depends also highly on your algorithm, here, it looks like an easy optimization. I think we could do more by using more advanced data structures, but it'd be unnecessary, as long as we have no reason to further optimize this part.

File: mangaki/mangaki/admin.py [hunk 10]
```diff
@@ -229,60 +230,99 @@ def make_nsfw(self, request, queryset):
...
+                                  "Mise à jour des tags effectuée pour une œuvre.")
+            else:
+                self.message_user(request,
+                                  "Mise à jour des tags effectuée pour {} œuvres.".format(nb_updates))
+            return None
+
+        # Check for works with missing AniDB AID
+        offending_works = []
+        if not all(work.anidb_aid for work in works):
+            offending_works = [work for work in works if not work.anidb_aid]
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: You could rewrite this as `filter(lambda w: not w.anidb_aid, works)` and combine it with the `map` (which, if I recall well, are some kind of wrapper over generator expression which is memory efficient).

Simple alternative: just use generator expression `(work for … …)` and it'll be the same theorically.

File: mangaki/mangaki/utils/anidb.py [hunk 11]
```diff
@@ -280,66 +283,60 @@ def _build_staff(self,
...
+        for tag_node in tags_soup.find_all('tag'):
+            tag_title = str(tag_node.find('name').string).strip()
+            tag_id = int(tag_node.get('id'))
+            tag_weight = int(tag_node.get('weight'))
+            tag_verified = tag_node.get('verified').lower() == 'true'
 
-                if tag_verified:
-                    tags[tag_title] = {'weight': tag_weight, 'anidb_tag_id': tag_id}
+            if tag_verified:
+                tags.append( AniDBTag(title=tag_title, weight=tag_weight, anidb_tag_id=tag_id) )
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: If my tool is not wrong, it seems like you added a white space after the `append(` and before the final closing parenthesis, those are unusual in Python, could you remove them?
If you want to make it more readable, an alternative would be:
```python
tags.append(
    AniDBTag(
        title=tag_title,
        weight=tag_weight,
        anidb_tag_id=tag_id
    )
)
```

File: mangaki/mangaki/utils/anidb.py [hunk 12]
```diff
@@ -280,66 +283,60 @@ def _build_staff(self,
...
-        added_tags = {key: anidb_tags[key] for key in added_tags_keys}
+        deleted_tags = tags_diff['deleted_tags']
+        added_tags = tags_diff['added_tags']
+        updated_tags = tags_diff['updated_tags']
+        kept_tags = tags_diff['kept_tags']
 
-        tags_id = [added_tags[title]["anidb_tag_id"] for title in added_tags]
-        existing_tags = Tag.objects.filter(anidb_tag_id__in=tags_id).all()
-        existing_tags_id = existing_tags.values_list('anidb_tag_id', flat=True)
+        tags_ids = [tag.anidb_tag_id for tag in deleted_tags+added_tags+updated_tags+kept_tags]
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: Could you add whitespaces between each `+`?

File: mangaki/mangaki/admin.py (hunk 9 above)
Comment (mrsalixor) [inline mangaki/mangaki/admin.py]: If I'm not mistaken, since tuples are hashable, I only need to store those tuples into a set rather than a list, and keep everything else as is ? I think it would be easier than creating a dict, since I don't use the associated values but only the keys.

Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: You're absolutely right and that is also a good solution, even better than a `dict` one I suppose (If I recall well, `set` is more memory efficient than `dict`!).

File: mangaki/mangaki/management/commands/anidb.py [hunk 13]
```diff
@@ -68,44 +68,46 @@ def handle(self, *args, **options):
...
             choice = input("Voulez-vous réaliser ces changements [y/n] : ")
             if choice == 'n':
                 print("\nOk, aucun changement ne va être fait")
             elif choice == 'y':
-                tags = deleted_tags
-                tags.update(added_tags)
-                tags.update(updated_tags)
-                tags.update(kept_tags)
-                client.update_tags(anime, tags)
+                all_tags = []
```
Comment (RaitoBezarius) [inline mangaki/mangaki/management/commands/anidb.py]: `all_tags = deleted_tags + added_tags + updated_tags + kept_tags` ?

Comment (mrsalixor) [inline mangaki/mangaki/management/commands/anidb.py]: Indeed ! 😅
